File: LeetCode/Math/JS/Leetcode464-canIWin.js
var canIWin = function (maxChoosableInteger, desiredTotal) {
    let [n, m] = [maxChoosableInteger, desiredTotal];
    let f = new Array(1 << n).fill(-1);
    if (Math.floor(n * (n + 1) / 2) < m) return false;
    let dp = function (x) {
        if (f[x] != -1) return f[x];
        let sum = 0;
        for (let i = 0; i < n; i++) {
            if ((x >> i) & 1) sum += i + 1;
        }
        for (let i = 0; i < n; i++) {
            if ((x >> i) & 1) continue;
            if (sum + i + 1 >= m) return f[x] = 1;
            if (!dp(x + (1 << i))) return f[x] = 1;
        }
        return f[x] = 0;
    }
    return dp(0);
};

File: LeetCode/Math/Rust/Leetcode464-canIWin.rs
impl Solution {
    pub fn can_i_win(max_choosable_integer: i32, desired_total: i32) -> bool {
        let (n, m) = (max_choosable_integer, desired_total);
        if n * (n + 1) / 2 < m {
            return false;
        }
        let mut f = vec![-1; 1 << n];
        if Solution::dp(&mut f, 0, n, m) == 1 {
            return true;
        }
        false
    }
    fn dp(f: &mut Vec<i8>, x: usize, n: i32, m: i32) -> i8 {
        if f[x] != -1 {
            return f[x];
        }
        let mut sum = 0;
        for i in 0..n {
            if (x >> i) & 1 == 1 {
                sum += i + 1;
            }
        }
        for i in 0..n {
            if (x >> i) & 1 == 1 {
                continue;
            }
            if sum + i + 1 >= m {
                f[x] = 1;
                return 1;
            }
            if Solution::dp(f, x + (1 << i), n, m) == 0 {
                f[x] = 1;
                return 1;
            }
        }
        f[x] = 0;
        0
    }
}
